### ISSUE TITLE
3D objects get jagged edges in GLES2

### ISSUE THREAD

**Issue by Skaruts** [godotengine/godot#33273]
3.2.alpha3 (same issue existed in previous versions, including 3.1.1 iirc -- and still existed a day or two ago in build `af4fd9de9`)
Windows 7 / ATI Radeon HD 4870

**This happens in GLES2 only** (at least in my case). In GLES3 all seems fine.

I get these weird glitches where the edges of **large** 3D objects are all jagged. This cube is 20x20x20, and the plane is 200x200:
![jagged](https://user-images.githubusercontent.com/6277234/68075646-67d43300-fda2-11e9-8f4d-4bc801e06ded.gif)

Doesn't matter if the size of the objects comes from scaling the MeshInstance or by adjusting the size of the Mesh itself.

I never noticed this until I bumped up the size of some assets by 10x, to try to fix a potential issue with physics. With the smaller ones everything seemed fine.

While materials don't solve the issue, setting one of the flags `Transparent`, `Unshaded` or `No Depth-Test`, makes the object render correctly, as is the case in the image below. Blue cube has transparency on, red cube doesn't. (Everything else in the materials is default.)
![jagged1](https://user-images.githubusercontent.com/6277234/68075684-f5b01e00-fda2-11e9-9e25-1efccb2bedf1.gif)

Amusingly, though, when I select the red cube, it suddenly renders perfectly fine. When I deselect it, it suddenly no longer does.

My first assumption is that it's related to depth testing. However, [in another issue](https://github.com/godotengine/godot/issues/32764#issuecomment-542140637) someone was suggesting tinkering with camera's `z-near`, and as I started bumping it up in small steps, the glitches started happening farther away from the camera (I had to zoom out further each time to make them happen). For the size of those cubes, setting it to something between 1.0 and 5.0 seems to make the glitches unnoticeable. So in a way that kinda "fixes" it.

(In the gif above I was using `z-far` 10000, but it didn't seem to me that `z-far` made any difference in this. I only made it big to zoom out at will. I noticed the issue when using z-near 0.1 and z-far 500, which are default values, if I'm not mistaken.)

**Steps to reproduce:**
Create a project, add a plane mesh and one or more cube meshes, and make their sizes somewhat big. As far as I could tell, the bigger, the more noticeable the glitches. Also, the more zoomed out, the more noticeable.

Sometimes the glitches don't start happening until I zoom out quite a bit and zoom back in.

**Minimal reproduction project:**
[Jagged-Edges-Test.zip](https://github.com/godotengine/godot/files/3800855/Jagged-Edges-Test.zip)

**Comment by clayjohn** [comment]
This looks like a driver bug to me. Especially with those odd shaped jagged edges appearing. 

Have you updated your GPU driver recently?

**Comment by Calinou** [comment]
I can't reproduce this on commit https://github.com/godotengine/godot/commit/ab2881bca1be5844dd17ae7b35922f851585a9ad with a GTX 1080 on Linux:

![image](https://user-images.githubusercontent.com/180032/68076713-be4c6c00-fdb8-11e9-8bc6-56bdfd85c3c4.png)

**Comment by Skaruts** [comment]
@Calinou Can you reproduce it in plain alpha3 (the public release)? 
(trying to figure out if the problem isn't on my end.)

@clayjohn My drivers have been uptodate since 2013. They haven't released any drivers for my card since.

**Comment by Calinou** [comment]
@Skaruts I can't reproduce this either on 3.2alpha3:

![image](https://user-images.githubusercontent.com/180032/68077635-46d20900-fdc7-11e9-9171-1f90103b86dd.png)

**Comment by clayjohn** [comment]
Also can't reproduce. Windows 10 NVidia GTX 1050. 

My guess is that this is a hardware/driver issue.

**Comment by logzero** [comment]
Driver is probably choosing 16bit depth precision by default, which is not that great for z-near 0.1 and z-far 500.

Anything beyond 465 maps to 65534 or 65535 for example:
https://www.opengl.org/archives/resources/faq/technical/depthbuffer.htm

**Comment by Skaruts** [comment]
Would making a material transparent or unshaded make the driver choose another precision depth?
Also, why doesn't it do that in GLES3?

**Comment by raphael10241024** [comment]
Reproduced on my Macbook pro, macOS Mojave 10.14.6
VideoCard:Radeon Pro 560X 4GB
<img width="615" alt="Screen Shot 2019-11-05 at 11 34 01 AM" src="https://user-images.githubusercontent.com/8491351/68176595-6e3ff600-ffc0-11e9-98d5-9241c86e38a8.png">

**Comment by logzero** [comment]
> Would making a material transparent or unshaded make the driver choose another precision depth?
> Also, why doesn't it do that in GLES3?

No idea. I just guessed it must have to do with the depth buffer, as it looks like depth testing artifacts. I am not sure if it applies to you, but older Catalyst drivers had an option to force z-depth.

From what I see in the code Godot seems to ask for 24 bit depth buffer for the framebuffer. It takes driver default for other renderbuffers  on desktop hardware.

You could try to figure out what the gpu is doing using apitrace or other 3d/opengl debugging tools.

**Comment by clayjohn** [comment]
@raphael10241024 Can you test this again from beta 6? There were some changes made to depth formats a few weeks after this issue was created. I am hoping it has been fixed, but I am still unable to reproduce in either 3.1.1 or 3.2 beta6

**Comment by Skaruts** [comment]
@clayjohn on my end it still does it, but only when objects are very distant, which makes it quite a bit less noticeable.
Tested in both beta6 and rc1.

**Comment by raphael10241024** [comment]
Looks bettrer than before , but still happens a little.
<img width="594" alt="Screen Shot 2020-01-19 at 11 42 03 PM" src="https://user-images.githubusercontent.com/8491351/72683792-6569b580-3b15-11ea-9e00-e4a77f5f0ac8.png">
